src/mosquito/runners/executor.cr

```diff
@@ -0,0 +1,89 @@
+module Mosquito::Runners
+  # An Executor is responsible for building Job classes with deserialized
+  # parameters and calling #run on them. It measures the time it takes to
+  # run a job and provides detailed log messages about the current status.
+  #
+  # Executor#deqeue_and_run_jobs is the entrypoint and shoud be treated as
+  # if it will return only after a relative eternity.
+  class Executor
+    include RunAtMost
+
+    Log = ::Log.for self
+
+    # How long a job config is persisted after success
+    property successful_job_ttl : Int32 { Mosquito.configuration.successful_job_ttl }
+
+    # How long a job config is persisted after failure
+    property failed_job_ttl : Int32 { Mosquito.configuration.failed_job_ttl }
+
+    getter queue_list : QueueList
+
+    def initialize(@queue_list)
+    end
+
+    def dequeue_and_run_jobs
+      queue_list.each do |q|
+        run_next_job q
+      end
+    end
+
+    def run_next_job(q : Queue)
+      job_run = q.dequeue
+      return unless job_run
+
+      Log.notice { "#{"Starting:".colorize.magenta} #{job_run} from #{q.name}" }
+
+      duration = Time.measure do
+        job_run.run
+      end.total_seconds
+
+      if job_run.succeeded?
+        Log.notice { "#{"Success:".colorize.green} #{job_run} finished and took #{time_with_units duration}" }
+        q.forget job_run
+        job_run.delete in: successful_job_ttl
+
+      else
+        message = String::Builder.new
+        message << "Failure: ".colorize.red
+        message << job_run
+        message << " failed, taking "
+        message << time_with_units duration
+        message << " and "
+
+        if job_run.rescheduleable?
+          next_execution = Time.utc + job_run.reschedule_interval
+          q.reschedule job_run, next_execution
+
+          message << "will run again".colorize.cyan
+          message << " in "
+          message << job_run.reschedule_interval
+          message << " (at "
+          message << next_execution
+          message << ")"
```

Comment:
Preferable to use `String.build` versus `String::Builder` directly.

Reply:
I usually do. I don't really see why the documentation recommends that. It's not inherently more readable, especially in this situation. Often it might be more readable, but not always.

In this situation I chose to use String::Builder.new ... #to_s instead because it requires one less named variable and doesn't further complicate the already complex logic for building the summary message. Honestly with how simple [String.build](https://github.com/crystal-lang/crystal/blob/879691b2e/src/string.cr#L278-L282) is, I don't even see what the benefit of using String.build actually is.